### PR TITLE
Ticket 4665 - Emoji Picker

### PR DESCRIPTION
**Pull request by bonespiked**
#### Summary
This creates an emoji picker for posting and replying to messages.

#### Ticket Link
https://github.com/mattermost/platform/issues/4665

#### Checklist
- [x] Has UI changes

Credit goes to bjoernr-de and harrison - I merely put it together at the end...

### CODE REVIEW

**Comment by jwilander** [comment]
Just played with it on the test server, this is fantastic guys! Feels really smooth and great to use

**Comment by bjoernr-de** [comment]
I'm missing my commits :stuck_out_tongue_winking_eye: - Just joking :wink: 

Ideally we should also consider putting the standard emojis into one big [sprite](http://www.w3schools.com/css/css_image_sprites.asp).
The amounts of requests on first opening of the emoji picker is still a bit of pain

**Comment by jasonblais** [comment]
Hey @bonespiked and @bjoernr-de, just writing notes on a separate Google doc, will share soon!

Overall, fantastic work 👍

**Comment by jasonblais** [comment]
Hey @bonespiked, @bjoernr-de, I added some feedback in the Google doc here: https://docs.google.com/document/d/1Pnvqs9JbBuYNKWvIr7b9NPH_Fc9SfX92TWqIeLS3ZLI/edit#

Please let me know if you have any questions, either here or on https://pre-release.mattermost.com/core?

None of the feedback I have is major, the emoji picker is working really smoothly! Fantastic work!

**Comment by jwilander** [comment]
Just a heads up the build is failing on some ESLint errors:
```
/home/ubuntu/workspace/mattermost-platform-pr/src/github.com/mattermost/platform/webapp/components/emoji_picker/components/emoji_picker_item.jsx
  42:15  error  'name' is assigned a value but never used  no-unused-vars

/home/ubuntu/workspace/mattermost-platform-pr/src/github.com/mattermost/platform/webapp/components/file_upload.jsx
  364:18  error  Expected indentation of 16 space characters but found 17  react/jsx-indent

✖ 2 problems (2 errors, 0 warnings)
```

**Comment by dgreene3p** [comment]
Ah - I had pushed a 'working' commit to my fork to gather assistance from some others, these _will_ be cleaned up.

**Comment by jasonblais** [comment]
Sweet! Great work @bonespiked!!

This is looking much better! Just adding some pieces of feedback, mostly minor nits on the UI.

### 1) Width of the picker

Propose we adjust the width of the picker so each row has 9 emoji? The picker is quite wide and feels big on the screen.

This might mean decreasing the whitespace in the header for the emoji icons.

If this doesn't work well, let me know, happy to consider 10 per row.

image

### 2) Emoji picker on mobile

Currently the emoji picker breaks on mobile. No action needed for now - let's see if the smaller width in 1) will help, but something we should think about.

image

### 3) File upload isn’t left-aligned with the message box on mobile view

image

### 4) Loading emoji on the picker

When you first open the emoji picker, the emojis load in (seemingly) random order.

Not sure what we can do so that the emojis would be pre-loaded before you open the picker for the first time? 

image

### 5) Floating search bar

Propose having the search bar "float" on the emoji picker. That way if you scroll to the bottom of the picker, the search bar will be visible at the top for easy searching.

image

### 6) Whitespace after clicking an icon in emoji picker header

 - click the "Activity" icon in emoji picker header
 - observed: not enough whitespace between the emoji header bar and the "Activity" title

image

 - expected: equal whitespace above and below the "Activity" title

image

### 7) Whitespace in the header

It looks like the whitespace between the emoji picker header icons isn't equal.

image

### 8) custom emoji for picker

Wide custom emoji don't have enough whitespace between the image and emoji text at the bottom of the picker

Observed: 
image

Expected: Same amount of whitespace as for regular emoji

image

### 9) Emoji picker border on right-hand sidebar

Thanks for changing the border colour for the emoji picker! Looks like it wasn't reflected on the right-hand sidebar though.

By the way, is it possible to use the same component for the center channel and right-hand sidebar, so that the format/design always matches?

image
^border in the center channel (correct)

image
^border in the right-hand sidebar

### Questions:
10) How much work is needed to add the emoji picker for emoji reactions?

11) Do we control the order of emojis (e.g. some of the smiley emoji appear at the bottom of the “People” category)?

image

### Future improvements (P2):

12) rather than showing X number of most recent emoji, show three rows of most recent emoji. That way if the emoji picker width changes, we'll continue to show three full rows of recent emoji
13) left/right/up/down arrows can be used to navigate through the emoji picker
14) add more flag emoji https://mattermost.atlassian.net/browse/PLT-5322
15) change emoji set neutral skin tone https://mattermost.atlassian.net/browse/PLT-5316

**Comment by dgreene3p** [comment]
I've made an update for items 1,2,6,8, and 9, and am working on the others.  I will try to finish / work on them over the weekend

**Comment by bonespiked** [comment]
I've finished items 1, 2, 6, 8, and 9, and will address the rest this weekend

**Comment by bonespiked** [comment]
I think I figured out what you were referring to with item number 3 - @jasonblais - can you please re-check the latest?

**Comment by jasonblais** [comment]
Thanks @bonespiked! Much appreciated!

Reviewed all feedback from the [Google doc](https://docs.google.com/document/d/1Pnvqs9JbBuYNKWvIr7b9NPH_Fc9SfX92TWqIeLS3ZLI/edit#) and the [GitHub post](https://github.com/mattermost/platform/pull/5157#issuecomment-278414278) above.

Most look fantastic, just minor notes on a couple of them

### 2) Emoji picker on mobile

I've queued an item to propose disabling the emoji picker on mobile. Otherwise I worry we'll need to spend A) lots of time perfecting the design for mobile that's B) unfamiliar to most mobile users anyways.

[You can see the post with more details on pre-release.mattermost.com](https://pre-release.mattermost.com/core/pl/w66en6prpfbwue5f7fmeqtkowy)

### 4) Loading emoji on the picker

> I will likely rework how it's rendered as part of emoji reaction, so I would like to postpone that one

That sounds good. How long do you think it'll be to add the emoji picker for reactions? You mentioned above that it would be "medium" work, but I'm not sure if that means a day, a week or a month ;) 

PS: I found its Jira ticket here: https://mattermost.atlassian.net/browse/PLT-2354

### 6) Whitespace after clicking an icon in emoji picker header

Sorry, the whitespace isn't quite equal above and below in the attached screenshot...just needs a bit more space above the header.

image

### 7) Whitespace in the header

Sorry, again, the whitespace doesn't seem quite equal. Maybe it's the icon sizes themselves since some of them are smaller?

image

### 8) Custom emoji for picker

This looks much better. Great work!! Minor nit: can we have the custom emoji vertically centered for the attached case (when the emoji is much wider)

PS: Wide custom emoji don't look great on the picker as the aspect ratio is lost. However, this is a separate issue from the picker and have a separate ticket to fix this (which should resolve it for the picker as well): https://mattermost.atlassian.net/browse/PLT-5358

image


### 11) Do we control the order of emojis?

> it's currently hard-coded in a datafile

Is it through this data file? https://github.com/mattermost/platform/blob/165ad0d4f791f8ae2109472d8a626d911fa368e0/webapp/utils/emoji.jsx

I noticed a lot of emoji are out of order actually and it's not directly related to this PR. I'll create a ticket for it so that someone can review the current emoji order and submit a PR for it separately -- not something that should block the progress for this PR.

-----

Couple of new ones:

### 16) Recent --> Recently Used

Sorry for not mentioning earlier: but can we change 'Recent' in the emoji header to 'Recently Used'?

It'll be more obvious what it's referring to.

### 17) The emoji picker is overlapping the message box

Probably a change caused by one of the recent commits

image

### 18) Emoji picker header icons

I realized the emoji picker header icons don't follow our [UX guidelines for button/icon states](https://docs.mattermost.com/developer/fx-guidelines.html#display-button-states).

Can you help:
  - add a tooltip for the header icons when hovering over them, e.g. `Recently Used`. (I think this used to work, but got removed when we removed the tooltip for emojis)
  - add an on_hover effect [as described in our guidelines](https://docs.mattermost.com/developer/fx-guidelines.html#display-button-states)

NOTE: The above two changes should ONLY apply to desktop, not the mobile browser or the mobile app (including the tablet). Button states are not used on mobile and tooltips on mobile have caused issues in the past.

----

PS: Also found an unrelated issue, filed a ticket here: https://mattermost.atlassian.net/browse/PLT-5477

**Comment by dgreene3p** [comment]
2- please let me know if you decide to skip mobile
4- I think it'll take me a week or two to find the time to knock it out
6- I did this intentionally, as the 'anchor' that moves, it puts the header in a good place.  More space above, and it looked like it was misaligned, reducing the space below made it look too crowded.
7- this is a visual artifact of the different widths/alignments/weights of the individual icons
8- I'll try to fix this
16 - I'll update
17- weird, but definitely needs fixing
18 - I'll fix this as well 

Thanks for the detailed feedback - I love/hate it 😄

**Comment by jasonblais** [comment]
2 - Let's disable the picker on mobile. 

However, we'll need to revisit the design when we add the picker for emoji reactions. The native emoji picker for mobile won't have custom emoji uploaded to Mattermost, which we'd want for reactions.

4 - Cool, that sounds good :) 

We might want to visit the proper rendering of emoji on the picker first, then move on to emoji reactions. What do you think?

6 - Okay cool, I think that's good then.

7 - Ah okay -- anything we can do for the plane (for `Travel`) and the lightbulb (`Symbols`)? Or is that size fixed? Otherwise the icons look okay..

8, 16, 17, 18: Thanks!

**Comment by dgreene3p** [comment]
I'm on 18 hover now - I should have a new push today

**Comment by alfredbez** [comment]
image
we should really [use a sprite](https://github.com/mattermost/platform/pull/5157#issuecomment-275262999)

**Comment by dgreene3p** [comment]
@alfredbez - agreed, it's slated for when I rework some of the display stuff for post reactions

**Comment by jasonblais** [comment]
@bonespiked I've retested and looks really good. Just minor notes:

### 17) The emoji picker is overlapping the message box

Looks like this is still an issue

**Center channel**
image

**Right-hand sidebar**
image

### 19) As discussed in the UX design channel, help add a sprite for the baseline emoji

If this requires too much work, just let me know. We can push this to a separate PR

### 20) As discussed in the UX design channel, add the picker behind feature flag

Add new setting to **Account Settings > Advanced > Preview pre-release features** with a checkbox
 - if selected, emoji icon with picker is shown for the message box
 - if unselected, emoji picker is not shown for the message box

The setting text should read `Enable new emoji picker in message input box`

**Comment by jasonblais** [comment]
Thanks @bonespiked, wonderful work. I've re-tested previous issues + the alignment issue and all works great. Just three notes

1 - I noticed that the emoji are different for sections not under "Recently Used" and "Custom". Probably related to the sprites, would you like to take a look?
  - emojis are bigger and a bit blurry
  - no more active state when hovering over the emoji (no shadow effect, and mouse cursor doesn't change)

image

2. I've reached out to a developer to help with the preview setting for emoji picker

3. I noticed if I logout of my account, the "Recently Used' list in the emoji picker is gone. Is this expected? I thought it worked before where the "Recently Used" emoji were saved between logins.

----

All else looks good! I've created tickets for future improvements (some tickets already existed, and most of these we discussed before):

P1:
1 - use emoji picker to react to individual posts (desktop only): https://mattermost.atlassian.net/browse/PLT-4854

P2:
2 - add more flag emoji https://mattermost.atlassian.net/browse/PLT-5322
3 - UX tweaks when when using mouse to hover over several emoji in emoji picker https://mattermost.atlassian.net/browse/PLT-5577
4 - review ordering of emoji on the picker https://mattermost.atlassian.net/browse/PLT-5575
5 - emoji picker for edit message dialog https://mattermost.atlassian.net/browse/PLT-5576

P3:
6 - left/right/up/down arrows can be used to navigate through the emoji picker: https://mattermost.atlassian.net/browse/PLT-5574
7 - add emoji picker for mobile / build custom emoji picker for mobile apps (no ticket for now, will look into it for the new mobile apps later)
8 - change emoji set to neutral skin tone https://mattermost.atlassian.net/browse/PLT-5316

**Comment by dgreene3p** [comment]
the difference in 'fuzzyness' is that the sprite version (all non-custom) is 20x20, where the original images are much larger, so they look better, including the 'fake' custom emojis that are included.  Recent emojis are pulled from the image files since they could be sprite based or custom-file based.

The recently used should be retained regardless - nothing in that code area was touched....

I bet the hover got clobbered when going to sprite... I'll look at that...

**Comment by bonespiked** [comment]
the difference in 'fuzzyness' is that the sprite version (all non-custom) is 20x20, where the original images are much larger, so they look better, including the 'fake' custom emojis that are included. Recent emojis are pulled from the image files since they could be sprite based or custom-file based.

The recently used should be retained regardless - nothing in that code area was touched....

I bet the hover got clobbered when going to sprite... I'll look at that...

**Comment by jasonblais** [comment]
Hmm, it does look like the 'recently used' disappears after logout/login..

Before logout:
image

After logout and subsequent login:
image

----

As for the fuzziness, anything we can do to correct that? The quality for 'fake' custom emojis is excellent, but the other ones aren't great..

----

Here are notes to help you set the emoji picker behind a preview feature setting. Hope this helps? Let me know if you have any questions about it

1. Add the pre-release setting to `webapp/utils/constants.jsx`
 - Look for the `PRE_RELEASE_FEATURES` object and add an `EMOJI_PICKER_PREVIEW` field. The `label` field will follow the pattern of the others, but I don't think the `description` field is used any more.

2. Go to `webapp/components/user_settings/user_settings_advanced.jsx` and add a case to the `renderFeatureLabel` function that will be used to render the label that'll show up for the setting
 - That should be all you need to add to this file, but I wouldn't be surprised if I missed something since the whole thing is a bit weird

3. In `webapp/components/create_post.jsx` and `webapp/components/create_comment.jsx`
Add a field to the state like `showEmojiPicker` which equals `Utils.isFeatureEnabled(PRE_RELEASE_FEATURES.EMOJI_PICKER_PREVIEW)`
 - Set that field in the constructor and in the `onPreferenceChange` method
 - Use that while rendering to tell whether or not to render the picker

**Comment by bonespiked** [comment]
I'll make the preview feature changes this weekend, as well as look into the dissappearing recently used.  

I'll also try to resize/rebuild the sprite with better looking images

**Comment by jasonblais** [comment]
Perfect, thanks @bonespiked!

Fantastic work with this feature, it's getting really close to completion :tada:

File: config/config.json [hunk 1]
```diff
@@ -21,7 +21,7 @@
         "EnablePostUsernameOverride": false,
         "EnablePostIconOverride": false,
         "EnableTesting": false,
-        "EnableDeveloper": false,
+        "EnableDeveloper": true,
```
Comment (jasonblais) [inline config/config.json]: Before queuing for dev review, we'll also want to remove these changes to config.json

File: webapp/package.json [hunk 2]
```diff
@@ -68,6 +68,7 @@
     "raw-loader": "0.5.1",
     "react-addons-test-utils": "15.4.2",
     "sass-loader": "4.1.1",
+    "react-outside-event": "^1.2.4",
```
Comment (jasonblais) [inline webapp/package.json]: License for https://github.com/gajus/react-outside-event looks good (BSD-3)

File: config/config.json (hunk 1 above)
Comment (bonespiked) [inline config/config.json]: changes undone

**Comment by bonespiked** [comment]
emoji picker is now controlled by preview setting (side note, preference names cannot be over 32 characters long total including 'feature_enabled_' prefix)
Sprite appearance is _much_ better - graphics improved, highlight hover working again, also cursor fixed.  I'm still looking into recently used before pushing commit.

**Comment by bonespiked** [comment]
Ok - tracked down logout behavior - recent emojis are stored in the browser localStorage, and on logout - GlobalActions.clientLogout() -> BrowserStore.clear() -> localStorage.clear().

This will need to be determined to be a dealbreaker or not, as persisting recent emojis to the db will be a bit of a rewrite.  I could write a hack to save the list before clearing localstorage, but that seems like a can open, worms everywhere path to tread down.

**Comment by jasonblais** [comment]
Reset the test server before realizing you hadn't pushed the commit yet :) 

> Ok - tracked down logout behavior - recent emojis are stored in the browser localStorage, and on logout - GlobalActions.clientLogout() -> BrowserStore.clear() -> localStorage.clear(). 

> This will need to be determined to be a dealbreaker or not, as persisting recent emojis to the db will be a bit of a rewrite. I could write a hack to save the list before clearing localstorage, but that seems like a can open, worms everywhere path to tread down.

Got it - this one is fine to wait until later and shouldn't block merging this PR. [I created a Jira ticket for us to track it](https://mattermost.atlassian.net/browse/PLT-5585), and added it to the above list of [future improvements](https://github.com/mattermost/platform/pull/5157#issuecomment-280785654)

**Comment by bonespiked** [comment]
Ok.  I'll get the update pushed this afternoon or evening...

**Comment by bonespiked** [comment]
updates pushed - once build/test passes, feel free to re-label the test server

**Comment by jasonblais** [comment]
Thanks @bonespiked, this is looking excellent 👍 

The most recent issues have been fixed. The preview flag is also working as expected.

I just found one small (sorry...) UI nit, which occurs when the preview flag is off (i.e. emoji picker is not enabled): the paperclip icon for file attachments is not vertically centered in the message box.

image

It's properly centered when the emoji picker is enabled.

image

File: config/config.json [hunk 3]
```diff
@@ -55,15 +55,15 @@
...
         "RestrictPublicChannelManagement": "all",
         "RestrictPrivateChannelManagement": "all",
+        "RestrictPublicChannelCreation": "all",
+        "RestrictPrivateChannelCreation": "all",
         "RestrictPublicChannelDeletion": "all",
         "RestrictPrivateChannelDeletion": "all",
         "UserStatusAwayTimeout": 300,
         "MaxChannelsPerTeam": 2000,
-        "MaxNotificationsPerChannel": 1000
+        "MaxNotificationsPerChannel": 1000000
```
Comment (jasonblais) [inline config/config.json]: can you also help revert the changes in config.json?

File: webapp/i18n/en.json [hunk 4]
```diff
@@ -1957,6 +1969,7 @@
   "user.settings.advance.slashCmd_autocmp": "Enable external application to offer slash command autocomplete",
   "user.settings.advance.title": "Advanced Settings",
   "user.settings.advance.webrtc_preview": "Enable the ability to make and receive one-on-one WebRTC calls",
+  "user.settings.advance.emojipicker": "Enable the emoji picker",
```
Comment (jasonblais) [inline webapp/i18n/en.json]: Last, propose having this text `Enable emoji picker in message input box`

File: config/config.json (hunk 3 above)
Comment (bonespiked) [inline config/config.json]: done.

File: webapp/i18n/en.json (hunk 4 above)
Comment (bonespiked) [inline webapp/i18n/en.json]: Done.

**Comment by bonespiked** [comment]
Made requested updates

File: config/config.json [hunk 5]
```diff
@@ -55,10 +55,10 @@
         "CustomDescriptionText": "",
         "RestrictDirectMessage": "any",
         "RestrictTeamInvite": "all",
-        "RestrictPublicChannelCreation": "all",
```
Comment (jasonblais) [inline config/config.json]: sorry, I'm still seeing changes to the config.json..

**Comment by jasonblais** [comment]
@asaadmahmood please help with CSS review? 

In the most recent commit (https://github.com/mattermost/platform/pull/5157/commits/97710f41829719b41f388f3bf961bbde72a4fbcf), @bonespiked changed the paperclip image to a font-awesome icon due to CSS issues. If this is unacceptable, can you help work with @bonespiked to make the PR work with the old paperclip image?

**Comment by jasonblais** [comment]
@asaadmahmood please help with review?

@bonespiked I think there isn't as much whitespace around the emoji icon as there used to be

image

I'm also seeing the edge of the emoji near the paperclip icon in mobile view

image

**Comment by asaadmahmood** [comment]
Similar to Jason, just add a `display: none` on `.post-create__container .post-create-body .icon__postcontent_picker` to hide the emoji icon on mobile.

Also make the attach icon and emoji icon similar in color, along with some increased spacing.
And, there's also this thing that when I hover near the buttons, both of them hover up, we should not include them in the same container. And have it only hover when we actually hover over the icon.

screen shot 2017-02-24 at 12 00 31 pm

**Comment by jasonblais** [comment]
@asaadmahmood would you be able to help take a look at your proposed changes above? We'll need @bonespiked's help focus on adding the picker for reactions

**Comment by asaadmahmood** [comment]
@jasonblais @bonespiked I've pushed a commit for the hover effect and spacing for the icons.

File: webapp/components/create_comment.jsx [hunk 6]
```diff
@@ -69,24 +73,85 @@ export default class CreateComment extends React.Component {
...
+            clickEvent.srcElement.className !== '' &&
+            clickEvent.srcElement.className.indexOf('emoji-rhs') === -1 &&
+            this.state.showEmojiPicker) {
+            this.setState({showEmojiPicker: !this.state.showEmojiPicker});
+        }
+    }
+
+    handleEmojiPickerClick() {
+        const threadHeight = $('#thread--root') ? $('#thread--root').outerHeight() : 0;
+        const messagesHeight = $('.post-right-comments-container') ? $('.post-right-comments-container').outerHeight() : 0;
```
Comment (jwilander) [inline webapp/components/create_comment.jsx]: Is there a way we can do this without using jQuery? We're really trying to get rid of it as it's very un-React-like

File: webapp/components/create_comment.jsx [hunk 7]
```diff
@@ -500,6 +573,20 @@ export default class CreateComment extends React.Component {
             addButtonClass += ' disabled';
         }
 
+        let emojiPicker = null;
+        if (this.state.showEmojiPicker) {
+            //TODO: get custom emojis
```
Comment (jwilander) [inline webapp/components/create_comment.jsx]: We shouldn't have any TODOs in code, we should make tickets for them. If this isn't going in for this PR @jasonblais can you work with @bonespiked to make a ticket for this? Then we can remove this TODO

File: webapp/components/create_post.jsx [hunk 8]
```diff
@@ -552,6 +610,19 @@ export default class CreatePost extends React.Component {
         if (!this.state.enableSendButton) {
             sendButtonClass += ' disabled';
         }
+        let emojiPicker = null;
+        if (this.state.showEmojiPicker) {
+            //TODO: get custom emojis
```
Comment (jwilander) [inline webapp/components/create_post.jsx]: Same as above

File: webapp/components/emoji_picker/emoji_picker.jsx [hunk 9]
```diff
@@ -0,0 +1,449 @@
...
+    flags: {
+        id: 'emoji_picker.flags',
+        defaultMessage: 'Flags'
+    },
+    custom: {
+        id: 'emoji_picker.custom',
+        defaultMessage: 'Custom'
+    }
+
+});
```
Comment (jwilander) [inline webapp/components/emoji_picker/emoji_picker.jsx]: We shouldn't use injectIntl or intlShape or defineMessages anymore. Can you use `utils.localizeMessage` and/or `FormattedMessage` instead?

File: webapp/components/emoji_picker/emoji_picker.jsx [hunk 10]
```diff
@@ -0,0 +1,449 @@
+// Copyright (c) 2016 Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+import React from 'react';
+
+import $ from 'jquery';
```
Comment (jwilander) [inline webapp/components/emoji_picker/emoji_picker.jsx]: Like my above comment, we are trying to avoid using jQuery. Can you replace any uses of it if at all possible?

File: webapp/package.json (hunk 2 above)
Comment (jwilander) [inline webapp/package.json]: Can we remove the `^` so it doesn't auto-update?

File: webapp/components/create_comment.jsx (hunk 6 above)
Comment (dgreene3p) [inline webapp/components/create_comment.jsx]: If you know a way to get the height of the RHS sections via react, I'm fine w/ it - it just has to dynamically position the picker depending on how tall the RHS section is.

File: webapp/components/create_comment.jsx (hunk 7 above)
Comment (dgreene3p) [inline webapp/components/create_comment.jsx]: agreed.  removed.

File: webapp/components/create_post.jsx (hunk 8 above)
Comment (dgreene3p) [inline webapp/components/create_post.jsx]: agreed. removed.

File: webapp/components/emoji_picker/emoji_picker.jsx (hunk 10 above)
Comment (dgreene3p) [inline webapp/components/emoji_picker/emoji_picker.jsx]: I'll try to find a way...

File: webapp/components/emoji_picker/emoji_picker.jsx (hunk 9 above)
Comment (dgreene3p) [inline webapp/components/emoji_picker/emoji_picker.jsx]: done.  Confirming now

File: webapp/components/create_comment.jsx (hunk 6 above)
Comment (jwilander) [inline webapp/components/create_comment.jsx]: Well I don't know of any React-friendly way to go about it but can we use vanilla JS here instead of jQuery. Select with `document.querySelector()` for example to get the element. Not sure how easy it is to get `outerHeight` without jQuery. If it becomes a monumental task to do this or the below work without jQuery I guess I'll have to be OK with it but I'd prefer to leave that as a last resort.

File: webapp/components/emoji_picker/emoji_picker.jsx (hunk 10 above)
Comment (jwilander) [inline webapp/components/emoji_picker/emoji_picker.jsx]: Thanks as I said above:

> If it becomes a monumental task to do this or the below work without jQuery I guess I'll have to be OK with it but I'd prefer to leave that as a last resort.

**Comment by dgreene3p** [comment]
I've made the changes sans the jQuery removal - still working on that...

**Comment by bonespiked** [comment]
Made updates sans jQuery removal - still working on that...

File: webapp/components/create_comment.jsx (hunk 6 above)
Comment (bonespiked) [inline webapp/components/create_comment.jsx]: was able to do this with plain js :
document.getElementById('thread--root').offsetHeight

**Comment by jwilander** [comment]
Chatted with @bonespiked, the majority of the uses of jQuery have been removed. Some are proving to be very painful to remove, and in interest of getting this long-running PR merged I am OK with leaving those jQuery uses in for now as long as we revisit them and remove later. I've created a [ticket here](https://mattermost.atlassian.net/browse/PLT-5671) for this.

Really nice job on this guys, appreciate all the work that's gone into it.

**Comment by jasonblais** [comment]
After re-testing, it was found that the emoji picker got broken on Firefox due to recent changes, so I have added the PM review label back. There's a small UI issue on IE11 and Edge that we'll fix at the same time.

Also chatted with @bonespiked and we'll cut the feature from 3.7 given today is the feature complete day. This helps ensure the feature meets the Mattermost quality bar and that the release is a success.

I'll queue it back to dev review once the new issues are resolved.

**Comment by jasonblais** [comment]
Anyone following the thread:

The issues affecting Firefox have now been fixed. Scrolling issue with IE11 has also been fixed.

There are two remaining fixes on the UI that will be resolved after which the PR is queued back to dev review. The issues are small so expecting the PM review phase to finish soon.

**Comment by hmhealey** [comment]
This looks great @bonespiked. One question about the sprite sheet, did you do that by hand or did you use some tool to generate it? I'd like to include anything you used in the documentation or along with the code used to generate `emoji.json` so that we can extend the sprite sheet when we add more emoji

**Comment by bonespiked** [comment]
I used:
http://spritegen.website-performance.org/

to generate the sprite as well as 98% of the CSS for _emojisprite.css.

**Comment by it33** [comment]
Looking really great! Excited to have this merged :) 

image

**Comment by jasonblais** [comment]
Queuing for dev review next, big thanks @bonespiked! 

Here's a list of further improvements and fixes for supporting full emoji reactions, not blocking the merge of this feature: https://mattermost.atlassian.net/browse/PLT-5896

**Comment by jasonblais** [comment]
Also requesting second review from @jwilander given latest code changes

File: webapp/components/create_post.jsx [hunk 11]
```diff
@@ -77,7 +82,10 @@ export default class CreatePost extends React.Component {
             fullWidthTextBox: PreferenceStore.get(Preferences.CATEGORY_DISPLAY_SETTINGS, Preferences.CHANNEL_DISPLAY_MODE, Preferences.CHANNEL_DISPLAY_MODE_DEFAULT) === Preferences.CHANNEL_DISPLAY_MODE_FULL_SCREEN,
             showTutorialTip: false,
             showPostDeletedModal: false,
-            enableSendButton: false
+            enableSendButton: false,
+            lastBlurAt: 0,
```
Comment (jwilander) [inline webapp/components/create_post.jsx]: I see `lastBlurAt` being added to the state here, but it's not used anywhere. Is that an accident? If so can we remove it?

File: webapp/components/emoji_picker/emoji_picker.jsx [hunk 12]
```diff
@@ -0,0 +1,410 @@
...
+        this.searchInput.focus();
+    }
+
+    onOutsideEvent = (event) => {
+        // Handle the event.
+        this.props.outsideClick(event);
+    }
+
+    handleCategoryClick(category) {
+        const items = ReactDOM.findDOMNode(this.refs.items);
```
Comment (jwilander) [inline webapp/components/emoji_picker/emoji_picker.jsx]: `ReactDOM.findDOMNode` is actually deprecated, you can just do `const items = this.refs.items;` instead. Can we fix the other places in this file we're using `ReactDOM.findDOMNode` as well?

File: webapp/components/emoji_picker/components/emoji_picker_preview.jsx [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+                name = emoji.aliases[0];
+                aliases = emoji.aliases;
+            }
+
+            return (
+                <div className='emoji-picker__preview'>
+                    <div className='emoji-picker__preview-image-box'>
+                        <img
+                            className='emoji-picker__preview-image'
+                            src={EmojiStore.getEmojiImageUrl(emoji)}
```
Comment (crspeller) [inline webapp/components/emoji_picker/components/emoji_picker_preview.jsx]: When I hover over each emoji I have to download that emoji from the server because this is using the specific emoji URL. This isn't a great experience on a slow connection. Can we change this to pull from the sprite sheet?

File: webapp/components/create_post.jsx (hunk 11 above)
Comment (bonespiked) [inline webapp/components/create_post.jsx]: it's used at line 190 and 505

File: webapp/components/emoji_picker/components/emoji_picker_preview.jsx (hunk 13 above)
Comment (bonespiked) [inline webapp/components/emoji_picker/components/emoji_picker_preview.jsx]: This method is to have one way to show images, as custom emoji are not part of the sprite sheet.

File: webapp/components/emoji_picker/emoji_picker.jsx (hunk 12 above)
Comment (bonespiked) [inline webapp/components/emoji_picker/emoji_picker.jsx]: I'll remove those I can - the ones that integrate jquery to handle scrolling I'm hesitant to remove - I can't get that to function w/o jquery (which removing jquery is an overall goal, I understand).

Comment (jwilander) [inline webapp/components/emoji_picker/emoji_picker.jsx]: I don't mean remove jquery (although I would like that but we've already discussed that) just replace `ReactDOM.findDOMNode(this.refs.items)` with `this.refs.items` because it's the same thing now. [If there's a ref to it, there's no need to use findDOMNode](https://facebook.github.io/react/docs/react-dom.html#finddomnode)

File: webapp/components/create_post.jsx (hunk 11 above)
Comment (jwilander) [inline webapp/components/create_post.jsx]: Those are using the `lastBlurAt` that is present directly on the component object itself, not the one in the state. The one in the state doesn't seem to be used anywhere

File: webapp/components/emoji_picker/components/emoji_picker_preview.jsx (hunk 13 above)
Comment (bonespiked) [inline webapp/components/emoji_picker/components/emoji_picker_preview.jsx]: I've made an update to use the sprite images for preview

File: webapp/components/create_post.jsx (hunk 11 above)
Comment (bonespiked) [inline webapp/components/create_post.jsx]: ah - gotcha - removed.

File: webapp/components/emoji_picker/emoji_picker.jsx (hunk 12 above)
Comment (bonespiked) [inline webapp/components/emoji_picker/emoji_picker.jsx]: removed in the two places non-jquery related.

**Comment by jasonblais** [comment]
@bonespiked Two issues, otherwise looks good

1) All emoji in the "Custom" section have a default preview when hovering over them rather than the actual image

image
image

2) When hovering over the emoji picker in the right-hand side, a vertical scrollbar appears for the thread causing the emoji picker to move whenever you hover on/off an emoji. I don't think it was there before.

image

**Comment by jwilander** [comment]
Congratulations guys, really nice work all around :1st_place_medal: 👍 Can't wait to start using it 😄
